README.md
RnaSeqSampleSize
============
* [Introduction](#Introduction)
* [User friendly web interface](#web)
* [Download and install](#download)
* [Example](#example)

<a name="Introduction"/>
# Introduction #
Sample size estimation is the most important issue in the design of RNA sequencing experiments. However, thousands of genes are quantified and tested for differential expression simultaneously in RNA-seq experiments. The false discovery rate for statistic tests should be controlled. At the same time, the thousands of genes have widely distributed read counts and dispersions, which were often estimated by experience or set at the most conservative values in previous sample size estimation methods. As a result, the estimated sample size will be inaccurate or over-estimated.

To solve these issues, we developed a sample size estimation method based on the distributions of gene read counts and dispersions from real data. Datasets from the user's preliminary experiments or the Cancer Genome Atlas (TCGA) can be used as reference. The read counts and their related dispersions will be selected randomly from the reference based on their distributions, and from that, the power and sample size will be estimated and summarized.

<a name="web"/>
# User friendly web interface #
A user friendly web interface for RnaSeqSampleSize package is provided at [CQS website](http://cqs.mc.vanderbilt.edu/shiny/RnaSeqSampleSize/). Most of the features in Example section can be found in this website.


<a name="download"/>
# Download and install #
You can download and install RnaSeqSampleSize package from [github](https://github.com/slzhao/RnaSeqSampleSize) by the following commands.

	#Running the following codes in your R
	library(devtools)
    install_github("slzhao/RnaSeqSampleSizeData")
    install_github("slzhao/RnaSeqSampleSize")

<a name="example"/>
# Example #
After you have installed RnaSeqSampleSize package. You can enter R and use following R codes to view documents and perform examples for it.
	
	#Load package
	library("RnaSeqSampleSize")
	#View vignette
	browseVignettes(package="RnaSeqSampleSize")
	#View help files
	?sample_size
	#Examples for sample size or power estimation by single read count and dispersion
	sample_size(power=0.8, f=0.01,rho=2, lambda0=5, phi0=0.5)
	est_power(n=63, rho=2, lambda0=5, phi0=0.5,f=0.01)
	#Examples for power estimation by prior real data, may use 3 minutes
	est_power_distribution(n=65,f=0.01,rho=2,distributionObject="TCGA_READ",repNumber=5)
	sample_size_distribution(power=0.918,f=0.01,rho=2,distributionObject="TCGA_READ",repNumber=5)
	#Examples for power curve generation
	result1<-est_power_curve(n=63, f=0.01, rho=2, lambda0=5, phi0=0.5)
	result2<-est_power_curve(n=63, f=0.05, rho=2, lambda0=5, phi0=0.5)
	plot_power_curve(list(result1,result2))



NEWS
CHANGES IN VERSION 0.99.8 2015-04-12
------------------------------------

BUG FIXES

    o Fix a bug in example of optimize_parameter function, which was caused by update of heatmap3 package. Thanks Dan.


CHANGES IN VERSION 0.99.7 2015-04-06
------------------------------------

BUG FIXES

    o Fix a bug in example of convertId function, which was caused by update of BioMart database. Thanks Dan.

CHANGES IN VERSION 0.99.5 2015-04-06
------------------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o Minor changes for genes with less than 1 read count;
 
CHANGES IN VERSION 0.99.4
-------------------------

NEW FEATURES

    o The user friendly web interface was improved

BUG FIXES

    o Bug fixed: Max sample size for distribution based sample size estimation;
    o Bug fixed: Keep consistent for all function names;

SIGNIFICANT USER-VISIBLE CHANGES

    o Other improvement based on the comments from Bioconductor reviewer;

CHANGES IN VERSION 0.99.3 2014-11-23
------------------------------------

SIGNIFICANT USER-VISIBLE CHANGES


    o New parameter: countFilterInRawDistribution, selectedGeneFilterByCount;
    o Other improvement based on the comments from Bioconductor reviewer;

CHANGES IN VERSION 0.99.2
-------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o The vignette was switched to a BiocStyle;
    o Other improvement based on the comments from Bioconductor reviewer;

CHANGES IN VERSION 0.99.1 2014-10-19
------------------------------------

SIGNIFICANT USER-VISIBLE CHANGES


    o Parameter was used to determine the power of genes below minAveCount;
    o Some recommendations from BiocCheck were improved;

CHANGES IN VERSION 0.99.0 2014-10-16
------------------------------------

SIGNIFICANT USER-VISIBLE CHANGES


    o Submit to Bioconductor.

CHANGES IN VERSION 0.5.0 2014-10-15
-----------------------------------

NEW FEATURES

    o The datasets from TCGA were moved to RnaSeqSampleSizeData package;
    o The examples were improved;
    o The vignette was generated;

CHANGES IN VERSION 0.4.0 2014-10-10
-----------------------------------

NEW FEATURES

    o Rcpp package was used to make some functions faster;
    o The package structure was improved;
    o The web interface was improved;

CHANGES IN VERSION 0.3.0 2014-08-01
-----------------------------------

NEW FEATURES

    o Beta approximate method was improved and used in most of the cases;
    o Read count and dispersion distribution estimation function was improved;
    o Estimation power or sample size by read count and dispersion distribution was improved.

RnaSeqSampleSize 0.2.0 2014-06-01
---------------------------------

NEW FEATURES

    o The function of estimation power or sample size by read count and dispersion distribution was imported and improved.

CHANGES IN VERSION 0.1.7 2014-03-30
-----------------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o An approximate estimation method was used to decrease the running time when average read count (lambda0) was larger than 20;

NEW FEATURES

    o A plotting power curve function plot_power_curve was provided;

CHANGES IN VERSION 0.1.6 2014-03-19
-----------------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o The code was improved to fit the web interface;

NEW FEATURES

    o The function est_power was provided for power estimation;

BUG FIXES

    o A bug was fixed;

CHANGES IN VERSION 0.1.5 2014-03-16
-----------------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o The C code for R function dnbinom was improved to decrease the running time greatly;
    o The powers for different N in estimating sample size were returned to prepare power curve;

CHANGES IN VERSION 0.1.4 2014-03-01
------------------------------------

NEW FEATURES

    o A user friendly web interface was provided at http://cqs.mc.vanderbilt.edu/shiny/RnaSeqSampleSize/;

CHANGES IN VERSION 0.1.3 2014-02-17
-----------------------------------


SIGNIFICANT USER-VISIBLE CHANGES

    o The algorithm was improved to decrease the memory usage;


CHANGES IN VERSION 0.1.2 2014-02-04
-----------------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o A more detail parameter to sample size table was provided;
    o The algorithm was improved to decrease the running time and memory usage;

CHANGES IN VERSION 0.1.1 2014-02-04
-----------------------------------


SIGNIFICANT USER-VISIBLE CHANGES

    o The algorithm was improved to decrease the running time greatly;

CHANGES IN VERSION 0.1.0 2014-01-24
-----------------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o A more detail parameter to sample size table was provided to make sample size estimation faster;
    o An estimated time to perform sample size estimation will be displayed;

BUG FIXES

    o A bug fixed when sample size N equals to 1;

CHANGES IN VERSION 0.0.1 2014-01-10
-----------------------------------

SIGNIFICANT USER-VISIBLE CHANGES


    o First version.


R/sampleSize.R
# TODO: Add comment
# 
# Author: zhaos
###############################################################################



##' sample_size
##' 
##' A function to estitamete the sample size for differential expression analysis of RNA-seq data.
##' 
##' A function to estitamete the sample size for differential expression analysis of RNA-seq data.
##' 
##' @param m Total number of genes for testing.
##' @param m1 Expected number of prognostic genes.
##' @param power Power to detecte prognostic genes.
##' @param f FDR level
##' @param w Ratio of normalization factors between two groups.
##' @param k Ratio of sample size between two groups.
##' @param rho minimum fold changes for prognostic genes between two groups.
##' @param lambda0 Average read counts for prognostic genes.
##' @param phi0 Dispersion for prognostic genes.
##' @param showMessage Logical. Display the message in the estimation process.
##' @param storeProcess Logical. Store the power and n in sample size or power estimation process.
##' @return Estimate sample size or a list including parameters and sample size in the process.
##' @export
##' @examples power<-0.8;rho<-2;lambda0<-5;phi0<-0.5;f<-0.01
##' sample_size(power=power, f=f,rho=rho, lambda0=lambda0, phi0=phi0)
sample_size<-function(power=0.8,m=20000, m1=200, f=0.1, k=1,w=1, rho=2, lambda0=5, phi0=1,showMessage=FALSE,storeProcess=FALSE){
	lambda0AppCut<-2000

	r1<-m1 * power
	beta<-1-power
	alpha_star<-r1*f/((m-m1)*(1-f))
	z_alpha<-qnorm(1-alpha_star/2, lower.tail=T)
	z_beta<-qnorm(power, lower.tail=TRUE)
	n_w<-( ( z_alpha + z_beta )^2* (1 + rho/w +phi0*lambda0*(1+rho^2)) )/ ( (rho-1)^2*lambda0 )
	
	start.point<-1
	end.point<-round(n_w)+30

	if (lambda0>=lambda0AppCut) {
		p1<-est_power_model(n=start.point,w=w,k=k, rho=rho, lambda0=lambda0, phi0=phi0, beta=beta, alpha=alpha_star)
	} else {
		p1<-est_power_root(n=start.point,w=w,k=k, rho=rho, lambda0=lambda0, phi0=phi0, beta=beta, alpha=alpha_star)
	}
	if (p1>0) {
		return(start.point)
	} else {
		step.power<-6
		if (lambda0>=lambda0AppCut) {
			n_Exact<-uniroot.integer(est_power_model, c(start.point, end.point), w=w,k=k, rho=rho, 
					lambda0=lambda0, phi0=phi0, beta=beta, alpha=alpha_star, pos.side=TRUE,
					step.up=TRUE, step.power=step.power,print.steps=showMessage)
		} else {
			n_Exact<-uniroot.integer(est_power_root, c(start.point, end.point), w=w,k=k, rho=rho, 
					lambda0=lambda0, phi0=phi0, beta=beta, alpha=alpha_star, pos.side=TRUE,
					step.up=TRUE, step.power=step.power,print.steps=showMessage) 
		}
		if (storeProcess) {
			n_Exact$process<-n_Exact$process[order(n_Exact$process[,1]),]
			n_Exact$process[,2]<-n_Exact$process[,2]+power
			n_Exact$parameters<-c(power,m, m1,  f, w, rho, lambda0, phi0,k)
			names(n_Exact$parameters)<-c("power","m", "m1",  "fdr", "w", "rho", "lambda0", "phi0","k")
		} else {
			n_Exact<-n_Exact$root
		}
		return(n_Exact)
	}
}

##' sample_size_distribution
##' 
##' A function to estitamete the sample size based on read counts and dispersion distribution in real data.
##' 
##' A function to estitamete the sample size based on read counts and dispersion distribution in real data.
##' 
##' @inheritParams sample_size
##' @inheritParams est_power_distribution
##' @return Estimate sample size or a list including parameters and sample size in the process.
##' @export
##' @examples \dontrun{
##' #Please note here the parameter repNumber was very small (5) to make the example code faster.
##' #We suggest repNumber should be at least set as 100 in real analysis.
##' sample_size_distribution(power=0.8,f=0.01,distributionObject="TCGA_READ",repNumber=5,showMessage=TRUE)
##' }
sample_size_distribution<-function(power=0.8,m=10000, m1=100, f=0.1, k=1,w=1, rho=2,showMessage=FALSE,storeProcess=FALSE,distributionObject,libSize,minAveCount=5,maxAveCount=2000,repNumber=100,dispersionDigits=1,seed=123,selectedGenes,pathway,species="hsa",countFilterInRawDistribution=TRUE,selectedGeneFilterByCount=FALSE){
	
	r1<-m1 * power
	beta<-1-power
	alpha_star<-r1*f/((m-m1)*(1-f))

	start.point<-1
	end.point<-800
	step.power<-5
	
	#process distribution
	temp<-selectDistribution(distributionObject=distributionObject,libSize=libSize,repNumber=repNumber,dispersionDigits=dispersionDigits,minAveCount=minAveCount,maxAveCount=maxAveCount,seed=seed,selectedGenes=selectedGenes,pathway=pathway,species=species,countFilterInRawDistribution=countFilterInRawDistribution,selectedGeneFilterByCount=selectedGeneFilterByCount)
	dispersionDistribution<-temp$selectedDispersion
	countDistribution<-temp$selectedCount	
	
	n_Exact<-uniroot.integer(est_power_distribution_sampleSize, c(start.point, end.point), w=w,k=k, rho=rho, 
				beta=beta, alpha=alpha_star, dispersionDistribution=dispersionDistribution,countDistribution=countDistribution,
				pos.side=TRUE,	step.up=TRUE, step.power=step.power,print.steps=showMessage)

	if (storeProcess) {
			n_Exact$process<-n_Exact$process[order(n_Exact$process[,1]),]
			n_Exact$process[,2]<-n_Exact$process[,2]+power
			n_Exact$parameters<-c(power,m, m1,  f, w, rho, k)
			names(n_Exact$parameters)<-c("power","m", "m1",  "fdr", "w", "rho","k")
	} else {
			n_Exact<-n_Exact$root
	}
	return(n_Exact)
}

est_power_distribution_sampleSize<-function(beta,...) {
		return(mean(est_power_distribution_root(...))-(1-beta))
}